clang-tools-extra/clang-tidy/bugprone/IncorrectEnableSharedFromThisCheck.cpp

```diff
@@ -0,0 +1,105 @@
+//===--- IncorrectEnableSharedFromThisCheck.cpp - clang-tidy --------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "IncorrectEnableSharedFromThisCheck.h"
+#include "clang/AST/ASTContext.h"
+#include "clang/AST/DeclCXX.h"
+#include "clang/AST/RecursiveASTVisitor.h"
+#include "clang/ASTMatchers/ASTMatchFinder.h"
+#include "clang/Basic/Specifiers.h"
+#include "llvm/ADT/SmallPtrSet.h"
+
+using namespace clang::ast_matchers;
+
+namespace clang::tidy::bugprone {
+
+void IncorrectEnableSharedFromThisCheck::registerMatchers(MatchFinder *Finder) {
+  Finder->addMatcher(translationUnitDecl(), this);
+}
+
+void IncorrectEnableSharedFromThisCheck::check(
+    const MatchFinder::MatchResult &Result) {
+
+  class Visitor : public RecursiveASTVisitor<Visitor> {
+    IncorrectEnableSharedFromThisCheck &Check;
+    llvm::SmallPtrSet<const CXXRecordDecl *, 16> EnableSharedClassSet;
+
+  public:
+    explicit Visitor(IncorrectEnableSharedFromThisCheck &Check)
+        : Check(Check) {}
+
+    bool VisitCXXRecordDecl(CXXRecordDecl *RDecl) {
+      if (!RDecl->hasDefinition()) {
+        return true;
+      }
+
+      if (isStdEnableSharedFromThis(RDecl))
+        EnableSharedClassSet.insert(RDecl->getCanonicalDecl());
+
+      for (const auto &Base : RDecl->bases()) {
+        const auto *BaseRecord =
+            Base.getType()->getAsCXXRecordDecl()->getCanonicalDecl();
+
+        if (EnableSharedClassSet.contains(BaseRecord) ||
+            isStdEnableSharedFromThis(BaseRecord)) {
+
+          if (isStdEnableSharedFromThis(BaseRecord) &&
+              Base.getAccessSpecifier() != clang::AS_public) {
+            const SourceRange ReplacementRange = Base.getSourceRange();
+            const std::string ReplacementString =
+                // Base.getType().getAsString() results in
+                // std::enable_shared_from_this<ClassName> or
+                // alias/typedefs of std::enable_shared_from_this<ClassName>
+                "public " + Base.getType().getAsString();
+            const FixItHint Hint = FixItHint::CreateReplacement(
+                ReplacementRange, ReplacementString);
+            Check.diag(
+                RDecl->getLocation(),
+                "this is not publicly inheriting from "
+                "std::enable_shared_from_this, will cause unintended behaviour "
+                "on shared_from_this. fix this by making it public inheritance",
+                DiagnosticIDs::Warning)
+                << Hint;
+          } else if (EnableSharedClassSet.contains(BaseRecord) &&
+                     Base.getAccessSpecifier() != clang::AS_public) {
+            // actually the same FixItHint, just different diag
```

Comment:
Then you could use the `%select` operator of the diagnostic builder to merge both cases:

```c++
diag(RDecl->getLocation(), 
     "%select{'std::enable_shared_from_this'|%1}0") << isStdEnableSharedFromThis(BaseRecord) << BaseRecord;
```
Note the `0` and `1` are indices to the `<<`-ed expressions.

Reply:
wonderful, didn't know this one existed. Glad to be able to get rid of the horrible code duplication/if nesting.